packages/application-generic/src/services/auth/auth.service.ts

```diff
@@ -90,7 +117,36 @@ export class AuthService {
     } else {
       if (authProvider === AuthProviderEnum.GITHUB) {
         const withoutUsername = user.tokens.find(
-          (i) => i.provider === AuthProviderEnum.GITHUB && !i.username && String(i.providerId) === String(profile.id)
+          (i) =>
+            i.provider === AuthProviderEnum.GITHUB &&
+            !i.username &&
+            String(i.providerId) === String(profile.id)
+        );
+
+        if (withoutUsername) {
+          await this.userRepository.update(
+            {
+              _id: user._id,
+              'tokens.providerId': profile.id,
+            },
+            {
+              $set: {
+                'tokens.$.username': profile.login,
+              },
+            }
+          );
+
+          user = await this.userRepository.findById(user._id);
+          if (!user) throw new ApiException('User not found');
+        }
+      }
+
+      if (authProvider === AuthProviderEnum.GOOGLE) {
+        const withoutUsername = user.tokens.find(
+          (token) =>
+            token.provider === AuthProviderEnum.GOOGLE &&
+            !token.username &&
+            String(token.providerId) === String(profile.id)
```

Comment:
This logic and the below code is duplicated between Google and GitHub. Can you please move it to a function?